_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,372 @@
+
+- Generates a random byte array of a specified size to serve as the payload for each message sent to the Kafka topic. Adds 65 to represent an uppercase letter in ASCII code (65 is 'A', 66 is 'B', and so on).
+- Ensures that the payload size is greater than zero and throws an `IllegalArgumentException` if not.
+
+**`retriable` method**
+
+- Determines whether to retry sending a message following a message sending error.
+- Returns `true` if the message sending process can be retried.
+- Returns `false` for null and specified exceptions, or those that do not implement the `RetriableException` interface.
+- Customizable to include other errors.
+
+    > Retries can lead to the possibility of duplicate messages being sent. 
+    > You can configure the producer to avoid duplicates using idempotency, as messages are delivered exactly once. 
+    > You can also use the `retries` configuration property to control how many times the producer will retry sending a message before giving up. 
+    > This setting affects how many times the `retriable` method may return `true` during a message send error. 
+    > For more information, see the [Strimzi documentation on ordered delivery](https://strimzi.io/docs/operators/latest/deploying#ordered_delivery).
+
+**`onCompletion` method**
+
+- Confirms successful message transmission and displays information about the message sent, including the topic, partition, and offset.
+- Prints an error message on exception. Appropriate action is taken based on whether it's a retriable or non-retriable error. If the error is retriable, the message sending process continues. If the error is non-retriable, a stack trace is printed and the producer is terminated.
+
+With the imported libraries, our constants, and these configuration properties and methods, the producer application can do all we set out to do.
+
+**Example producer application**
+```java
+import java.util.Properties;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.LongSerializer;
+
+public class Producer implements Callback {
+    // Constants for configuration
+    private static final Random RND = new Random(0);
+    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
+    private static final String TOPIC_NAME = "my-topic";
+    private static final long NUM_MESSAGES = 100;
+    private static final int MESSAGE_SIZE_BYTES = 100;
+    private static final long PROCESSING_DELAY_MS = 0L;
```

Comment:
Is it intentional using 0 delay? With this value the producer would work as expected but I would like to make a broader point ...

The `producer.send` is asynchronous and it just puts the data in the buffer to make a batch. It will be sent based on  `batch.size` (default 16384) and `linger.ms` (default 0). Due to default 0 `linger.ms` each batch with one message (just 100 bytes) will be sent immediately.
If the reader changes `PROCESSING_DELAY_MS` with an higher value, i.e. 1000 ms, the code will still work as expected and sending each message every 1000 ms.
Things become complicated if the reader starts playing with above `batch.size` and `linger.ms`.
For example, setting `PROCESSING_DELAY_MS = 100` but `linger.ms = 500` will have the result of not sending each message every 100 ms. But each iteration in the loop would give time to the internal producer buffer to be filled with 4-5 messages (up to the 500ms) and then all sent together.
In conclusion, we should not make the blog post much complicated but at least one note for the user to be aware of how `batch.size` and `linger.ms` can have an influence on the expected `PROCESSING_DELAY_MS`?

Also, when running the example the client will send 100 messages in a really short time. Maybe we could have less messages to send and an higher delay between each message to give time to the user seeing the producer sending messages (i.e. 10 or 50 messages every 1000 ms?).

Reply:
I've changed the defaults to sending 50 messages every 1000 ms and added a bit more context. 
We have a section on tuning later which discusses linger and batch size.

Reply:
FYI @fvaleri^

Reply:
yeah but you just mention `Use the `linger.ms` and `batch.size` configuration properties to batch more messages into a single produce request for higher throughput. ` ... there is no explanation about what I said here.

Reply:
Thanks @ppatierno. I've tried to keep it minimal, as I don't want the post to get so dense it's not easily digestible.

How's this:

The `producer.send` method is asynchronous and buffers messages for batching.
Use the `linger.ms` and `batch.size` configuration properties to batch more messages into a single produce request for higher throughput. 
Modifying these properties impacts the message sending rate in relation to the `PROCESSING_DELAY_MS` setting.
For example, if you set `PROCESSING_DELAY_MS` to 1000 ms and `linger.ms` to 5000 ms, messages are batched and sent every 5000 ms (as specified by `linger.ms`).